llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+        Graph[&F].insert(PotentialCallees.begin(), PotentialCallees.end());
+      }
+    }
+
+    // And every global variable (but their handling is a bit simpler)
+    for (const auto &GV : M.globals())
+      for (const Value *U : GV.users())
+        addUserToGraphRecursively(cast<const User>(U), &GV);
+  }
+
+  iterator_range<GlobalSet::const_iterator>
+  dependencies(const GlobalValue *Val) const {
+    auto It = Graph.find(Val);
+    return (It == Graph.end())
+               ? make_range(EmptySet.begin(), EmptySet.end())
+               : make_range(It->second.begin(), It->second.end());
+  }
+
+private:
+  void addUserToGraphRecursively(const User *Root, const GlobalValue *V) {
+    SmallVector<const User *, 8> WorkList;
+    WorkList.push_back(Root);
+
+    while (!WorkList.empty()) {
+      const User *U = WorkList.pop_back_val();
+      if (const auto *I = dyn_cast<const Instruction>(U)) {
+        const auto *UFunc = I->getFunction();
+        Graph[UFunc].insert(V);
+      } else if (isa<const Constant>(U)) {
+        if (const auto *GV = dyn_cast<const GlobalVariable>(U))
+          Graph[GV].insert(V);
+        // This could be a global variable or some constant expression (like
+        // bitcast or gep). We trace users of this constant further to reach
+        // global objects they are used by and add them to the graph.
+        for (const auto *UU : U->users())
+          WorkList.push_back(UU);
+      } else
+        llvm_unreachable("Unhandled type of function user");
+    }
+  }
+
+  DenseMap<const GlobalValue *, GlobalSet> Graph;
+  SmallPtrSet<const GlobalValue *, 1> EmptySet;
+};
+
+void collectFunctionsAndGlobalVariablesToExtract(
+    SetVector<const GlobalValue *> &GVs, const Module &M,
+    const EntryPointGroup &ModuleEntryPoints, const DependencyGraph &Deps,
+    const std::function<bool(const Function *)> &IncludeFunctionPredicate =
+        nullptr) {
+  // We start with module entry points
+  for (const auto *F : ModuleEntryPoints.Functions)
+    GVs.insert(F);
+
+  // Non-discardable global variables are also include into the initial set
+  for (const auto &GV : M.globals()) {
+    if (!GV.isDiscardableIfUnused())
+      GVs.insert(&GV);
+  }
+
+  // GVs has SetVector type. This type inserts a value only if it is not yet
+  // present there. So, recursion is not expected here.
+  size_t Idx = 0;
+  while (Idx < GVs.size()) {
+    const auto *Obj = GVs[Idx++];
+
+    for (const GlobalValue *Dep : Deps.dependencies(Obj)) {
+      if (const auto *Func = dyn_cast<const Function>(Dep)) {
+        if (Func->isDeclaration())
+          continue;
+
+        // Functions can be additionally filtered
+        if (!IncludeFunctionPredicate || IncludeFunctionPredicate(Func))
+          GVs.insert(Func);
+      } else {
+        // Global variables are added unconditionally
+        GVs.insert(Dep);
+      }
+    }
+  }
+}
+
+// Check "spirv.ExecutionMode" named metadata in the module and remove nodes
```

Comment:
How is that metadata used and what is it relation to SYCL? I can probably guess that some kernel properties are propagated through it and we invalidate it during device code split by copying the metadata, but not copying all the kernels. I feel like we should be more elaborate about what's going on in the code in the upstream.

Reply:
This is removed.